src/diff.rs

```diff
@@ -0,0 +1,524 @@
+        };
+      }
+    }
+    super::native::get_sad(plane_org, plane_ref, blk_w, blk_h, bit_depth)
+  }
+}
+
+mod native {
+  use crate::tiling::*;
+  use crate::util::*;
+
+  #[inline(always)]
+  pub fn get_sad<T: Pixel>(
+    plane_org: &PlaneRegion<'_, T>, plane_ref: &PlaneRegion<'_, T>,
+    blk_w: usize, blk_h: usize, _bit_depth: usize
+  ) -> u32 {
+    let mut sum = 0 as u32;
+
+    for (slice_org, slice_ref) in
+      plane_org.rows_iter().take(blk_h).zip(plane_ref.rows_iter())
+    {
+      sum += slice_org
+        .iter()
+        .take(blk_w)
+        .zip(slice_ref)
+        .map(|(&a, &b)| (i32::cast_from(a) - i32::cast_from(b)).abs() as u32)
+        .sum::<u32>();
+    }
+
+    sum
+  }
+
+  #[inline(always)]
+  fn butterfly(a: i32, b: i32) -> (i32, i32) {
+    ((a + b), (a - b))
+  }
+
+  #[inline(always)]
+  fn hadamard4_1d(data: &mut [i32], n: usize, stride0: usize, stride1: usize) {
+    for i in 0..n {
+      let sub: &mut [i32] = &mut data[i * stride0..];
+      let (a0, a1) = butterfly(sub[0 * stride1], sub[1 * stride1]);
+      let (a2, a3) = butterfly(sub[2 * stride1], sub[3 * stride1]);
+      let (b0, b2) = butterfly(a0, a2);
+      let (b1, b3) = butterfly(a1, a3);
+      sub[0 * stride1] = b0;
+      sub[1 * stride1] = b1;
+      sub[2 * stride1] = b2;
+      sub[3 * stride1] = b3;
+    }
+  }
+
+  #[inline(always)]
+  fn hadamard8_1d(data: &mut [i32], n: usize, stride0: usize, stride1: usize) {
+    for i in 0..n {
+      let sub: &mut [i32] = &mut data[i * stride0..];
+
+      let (a0, a1) = butterfly(sub[0 * stride1], sub[1 * stride1]);
+      let (a2, a3) = butterfly(sub[2 * stride1], sub[3 * stride1]);
+      let (a4, a5) = butterfly(sub[4 * stride1], sub[5 * stride1]);
+      let (a6, a7) = butterfly(sub[6 * stride1], sub[7 * stride1]);
+
+      let (b0, b2) = butterfly(a0, a2);
+      let (b1, b3) = butterfly(a1, a3);
+      let (b4, b6) = butterfly(a4, a6);
+      let (b5, b7) = butterfly(a5, a7);
+
+      let (c0, c4) = butterfly(b0, b4);
+      let (c1, c5) = butterfly(b1, b5);
+      let (c2, c6) = butterfly(b2, b6);
+      let (c3, c7) = butterfly(b3, b7);
+
+      sub[0 * stride1] = c0;
+      sub[1 * stride1] = c1;
+      sub[2 * stride1] = c2;
+      sub[3 * stride1] = c3;
+      sub[4 * stride1] = c4;
+      sub[5 * stride1] = c5;
+      sub[6 * stride1] = c6;
+      sub[7 * stride1] = c7;
+    }
+  }
+
+  #[inline(always)]
+  fn hadamard2d(data: &mut [i32], (w, h): (usize, usize)) {
+    /*Vertical transform.*/
+    let vert_func = if h == 4 { hadamard4_1d } else { hadamard8_1d };
+    vert_func(data, w, 1, h);
+    /*Horizontal transform.*/
+    let horz_func = if w == 4 { hadamard4_1d } else { hadamard8_1d };
+    horz_func(data, h, w, 1);
+  }
+
+  fn hadamard4x4(data: &mut [i32]) {
+    hadamard2d(data, (4, 4));
+  }
+
+  fn hadamard8x8(data: &mut [i32]) {
+    hadamard2d(data, (8, 8));
+  }
+
+  #[allow(unused)]
+  #[inline(always)]
+  pub fn get_satd<T: Pixel>(
```

Comment:
Needs to be documented about what it is (4x4 or 8x8 hadamard, larger sizes tiled)